File: Reorder List.py
# Definition for singly-linked list.
def traverse(head):
    if head==None:
        return
    current=head
    while current!=None:
        print current.val,"->"
        current=current.next
def reverse(head):
    if head==None or head.next==None:
        return head
    cur=head.next
    head.next=None
    while cur!=None:
        t=cur.next
        cur.next=head
        head=cur
        cur=t
    return head
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return nothing

    def reorderList(self, head):
        if head==None:
            return head
        current=head
        l=0
        while current!=None:
            l+=1
            current=current.next
        halfL=l/2
        current=head
        while halfL>0:
            current=current.next
            halfL-=1
        half=current
        # print half.val
        reversedHalf=reverse(half.next)
        # traverse(reversedHalf)
        half.next=None
        current=head
        halfCurrent=reversedHalf
        while halfCurrent!=None:
            temp=current.next
            halfTemp=halfCurrent.next
            current.next=halfCurrent
            halfCurrent.next=temp
            current=temp
            halfCurrent=halfTemp
        return head
        
        
        # O(n^2)
        # cur=half
        # current=head
        # while current!=half:
        #     if cur.next==None:
        #         return head
        #     while cur.next.next!=None:
        #         cur=cur.next
        #     temp=current.next
        #     current.next=cur.next
        #     cur.next.next=temp
        #     current=temp
        #     cur.next=None
        #     cur=half
        # return head
        
    
        
        
        
l1=ListNode(1)
l2=ListNode(2)
l3=ListNode(3)
l4=ListNode(4)
l5=ListNode(5)
l6=ListNode(6)
# l7=ListNode(7)
l1.next=l2
l2.next=l3
l3.next=l4
l4.next=l5
l5.next=l6
# l6.next=l7
s=Solution()
a=s.reorderList(l1)
traverse(a)

File: java/ReorderList.java
 class ListNode {
     int val;
     ListNode next;
     ListNode(int x) {
		 val = x;
         next = null;
     }
 }
public class ReorderList{
	public void traverse(ListNode head){
		ListNode cur=head;
		while(cur!=null){
			System.out.print(cur.val);
			System.out.print("->");
			cur=cur.next;
		}
	}
	public ListNode reverse(ListNode head){
		if(head==null || head.next==null)
			return head;
		ListNode cur=head.next;
		head.next=null;
		while(cur!=null){
			ListNode t=cur.next;
			cur.next=head;
			head=cur;
			cur=t;
		}
		return head;
	}

    public void reorderList(ListNode head) {
		if(head==null)
			return;
        ListNode current=head;
		int len=0;
		while(current!=null){
			len++;
			current=current.next;
		}
		current=head;
		int half=len/2+1;
		while(half>1){
			current=current.next;
			half--;
		}
		ListNode x= current.next;
		current.next=null;

		x=reverse(x);
		current=head;
		while(x!=null){
			ListNode t=current.next;
			ListNode y=x.next;
			current.next=x;
			x.next=t;
			current=t;
			x=y;
		}
		
    }
	public static void main(String args[]){
		ListNode l1=new ListNode(1);
		ListNode l2=new ListNode(2);
		ListNode l3=new ListNode(3);
		ListNode l4=new ListNode(4);
		ListNode l5=new ListNode(5);
		// ListNode l6=new ListNode(6);
		l1.next=l2;
		l2.next=l3;
		l3.next=l4;
		l4.next=l5;
		// l5.next=l6;
		ReorderList rl=new ReorderList();
		rl.reorderList(l1);
		rl.traverse(l1);
	}
}